calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+        vec1.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+        for (k, hs) in vec1.into_iter() {
+            write!(f, "Node: {:?} --", k)?;
+            let mut vec = hs.into_iter().collect::<Vec<_>>();
+            vec.sort_unstable();
+            writeln!(f, " Dominators: {:?}", vec)?;
+        }
+        write!(f, "}}")
+    }
+}
+
+// Given a control stmt, returns Some(val) where val is the value of attribute s
+// of stmt. Returns None if no s attribute exists.
+fn get_attr(stmt: &ir::Control, s: &str) -> Option<u64> {
+    stmt.get_attributes().and_then(|atts| atts.get(s)).copied()
+}
+
+/// Caleb Note: This is a copy+ paste from the tdcc pass that I edited slightly.
+/// We should unify at some point.
+/// Adds the @NODE_ID attribute to all control stmts except emtpy ones.
+/// Also, for If stmts, instead of an @NODE_ID, it gets a beginning and end
+/// id.
+///
+/// ## Example:
+/// ```
+/// seq { A; if cond {X} else{Y}; par { C; D; }; E }
+/// ```
+///
+/// gets the labels:
+///
+/// ```
+/// @NODE_ID(0)seq {
+///   @NODE_ID(1) A;
+///   @BEGIN_ID(2) @END_ID(5) if cond {
+///     @NODE_ID(3) X
+///   }
+///   else{
+///     @NODE_ID(4) Y
+///   }
+///   @NODE_ID(6) par {
+///     @NODE_ID(7) C;
+///     @NODE_ID(8) D;
+///   }
+///   @NODE_ID(9) E;
+/// }
+///
+/// These identifiers are used by the compilation methods [calculate_states_recur]
+/// and [control_exits]
+fn compute_unique_ids(con: &mut ir::Control, mut cur_state: u64) -> u64 {
+    match con {
+        ir::Control::Enable(ir::Enable { attributes, .. })
+        | ir::Control::Invoke(ir::Invoke { attributes, .. }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state + 1
+        }
+        ir::Control::Par(ir::Par {
+            stmts, attributes, ..
+        })
+        | ir::Control::Seq(ir::Seq {
+            stmts, attributes, ..
+        }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            stmts.iter_mut().for_each(|stmt| {
+                let new_state = compute_unique_ids(stmt, cur_state);
+                cur_state = new_state;
+            });
+            cur_state
+        }
+        ir::Control::If(ir::If {
+            tbranch,
+            fbranch,
+            attributes,
+            ..
+        }) => {
+            attributes.insert(BEGIN_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(tbranch, cur_state);
+            cur_state = compute_unique_ids(fbranch, cur_state);
+            attributes.insert(END_ID, cur_state);
+            cur_state + 1
+        }
+        ir::Control::While(ir::While {
+            body, attributes, ..
+        }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(body, cur_state);
+            cur_state
+        }
+        ir::Control::Empty(_) => cur_state,
+    }
+}
+
+impl DominatorMap {
```

Comment:
A bunch of these functions, like `get_id` and `get_attr` don't seem to be related to a DominatorMap in particular. In that case, we should move them into the file scope and use them there. In general, functions that act upon a DominatorMap reference or construct a DominatorMap should be inside the `impl`.